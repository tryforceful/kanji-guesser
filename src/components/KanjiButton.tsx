import { IonButton } from "@ionic/react"
import classNames from "classnames"
import React from "react"
import { KanjiCharacter } from "../data/QuizData"
import { UserChoice } from "./Quizzard"

interface Props {
  thisButtonsKanji: KanjiCharacter
  correctChoice: KanjiCharacter
  userChoice: UserChoice
  onClick: (chosenKanji: KanjiCharacter) => void
}

const KanjiButton: React.FC<Props> = ({
  thisButtonsKanji,
  correctChoice,
  userChoice,
  onClick: emitButtonPress
}) => {
  return (
    <IonButton
      className={classNames([
        "kanji-choice",
        {
          selected: thisButtonsKanji === userChoice
        }
      ])}
      strong={false}
      fill="solid"
      // {!userChoice ? "outline" : "solid"}
      disabled={!!userChoice}
      color={!userChoice ? "primary" : thisButtonsKanji === correctChoice ? "success" : "danger"}
      size="large"
      onClick={() => emitButtonPress(thisButtonsKanji)}
    >
      {thisButtonsKanji}
    </IonButton>
  )
}

export default KanjiButton
