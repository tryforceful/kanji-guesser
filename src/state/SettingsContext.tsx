// inspired by https://kentcdodds.com/blog/application-state-management-with-react

import React, { useState } from "react"

export enum KanaDisplay {
  Kana = "かな",
  Romaji = "Romaji"
}

export enum QuizDifficulty {
  Easy = 4,
  Medium = 8,
  Hard = 12
}

export interface SettingsObj {
  darkThemeOn: boolean
  kanaDisplayChoice: KanaDisplay
  quizDifficulty: QuizDifficulty
}

export const initialSettings: SettingsObj = {
  darkThemeOn: false,
  kanaDisplayChoice: KanaDisplay.Kana,
  quizDifficulty: QuizDifficulty.Medium
}

export type SettingsContext = [SettingsObj, React.Dispatch<React.SetStateAction<SettingsObj>>]

export const SettingsContext: React.Context<SettingsContext> = React.createContext(
  // This one will not be used unless the context is refrenced while providerless
  //useState(initialSettings)
  (undefined as unknown) as SettingsContext
)

function useSettings(): SettingsContext {
  const context = React.useContext(SettingsContext)
  if (!context) throw new Error("useSettings must be used within a SettingsContext.Provider")
  return context
}

const SettingsProvider: React.FC<{}> = props => {
  //const [settings, setSettings] = useState(initialSettings)
  //const value = React.useMemo(() => [settings, setSettings], [settings])
  return <SettingsContext.Provider value={useState(initialSettings)} {...props} />
}

export { SettingsProvider, useSettings }
