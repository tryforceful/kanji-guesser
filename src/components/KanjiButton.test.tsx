import React from 'react'
import KanjiButton from './KanjiButton';
import { SettingsProvider } from '../state/SettingsContext';

//import renderer from 'react-test-renderer'
import { render } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect'

// import ReactDOM from "react-dom";
// import { act } from "react-dom/test-utils";

describe("Kanji Button tests", function() {

  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation(query => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  })

  // let container;
  // beforeEach(() => {
  //   container = document.createElement("div");
  //   document.body.appendChild(container);
  // });

  // afterEach(() => {
  //   document.body.removeChild(container);
  //   container = null;
  // });
  
  test('renders without crashing', () => {
    const { baseElement, unmount } = render(
    <SettingsProvider>
      <KanjiButton
        thisButtonsKanji="信"
        correctChoice="b"
        userChoice={null}
        onClick={()=>{}}
      />
    </SettingsProvider>
    );
    expect( baseElement ).toBeDefined();
    unmount();
  });

  test('button changes when user makes a choice', () => {

    //let component: renderer.ReactTestRenderer;
    const { baseElement, unmount, debug, rerender, getByText } = render(
        <SettingsProvider>
          <KanjiButton
            thisButtonsKanji="信"
            correctChoice="b"
            userChoice={null}
            onClick={()=>{}}
          />
        </SettingsProvider>
        );
    debug();

    expect(baseElement).toBeInTheDocument();

      // toJSON, toTree, update, unmount, getInstance, root
    expect(baseElement).toMatchSnapshot();

    // incorrect choice
    rerender(
        <SettingsProvider>
          <KanjiButton
            thisButtonsKanji="信"
            correctChoice="b"
            userChoice={"a"}
            onClick={()=>{}}
          />
        </SettingsProvider>
    );

    expect(baseElement).toMatchSnapshot();
    let _button = getByText("信")
    expect(_button).not.toHaveClass("selected");
    expect(_button).toHaveAttribute("color", "danger");

    // correct choice
    rerender(
        <SettingsProvider>
          <KanjiButton
            thisButtonsKanji="b"
            correctChoice="b"
            userChoice={"a"}
            onClick={()=>{}}
          />
        </SettingsProvider>
    );
    expect(baseElement).toMatchSnapshot();
    _button = getByText("b")
    expect(_button).not.toHaveClass("selected");
    expect(_button).toHaveAttribute("color", "success");

    unmount();
  });
})