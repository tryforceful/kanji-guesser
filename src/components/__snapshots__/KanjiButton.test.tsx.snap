// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kanji Button tests button changes when user makes a choice 1`] = `
<body>
  <div>
    <ion-button
      class="kanji-choice desktop"
      color="primary"
      fill="solid"
    >
      信
    </ion-button>
  </div>
</body>
`;

exports[`Kanji Button tests button changes when user makes a choice 2`] = `
<body>
  <div>
    <ion-button
      class="kanji-choice desktop"
      color="danger"
      fill="solid"
    >
      信
    </ion-button>
  </div>
</body>
`;

exports[`Kanji Button tests button changes when user makes a choice 3`] = `
<body>
  <div>
    <ion-button
      class="kanji-choice desktop"
      color="success"
      fill="solid"
    >
      b
    </ion-button>
  </div>
</body>
`;
